C/CUTEst/build_tarballs.jl

```diff
@@ -0,0 +1,128 @@
+
+# build SIFDecode
+cd $SIFDECODE
+
+if [[ "${target}" == *-linux* || "${target}" == *-freebsd* ]]; then
+  echo "5" > sifdecode.opts   # PC
+  echo "2" >> sifdecode.opts  # Linux
+  echo "5" >> sifdecode.opts  # gfortran
+elif [[ "${target}" == *-apple* ]]; then
+  echo "13" > sifdecode.opts  # macOS
+  echo "2" >> sifdecode.opts  # gfortran
+elif [[ "${target}" == *-mingw* ]]; then
+  cd $ARCHDEFS
+  cp compiler.pc64.msys2.gfo compiler.pc.mgw.gfo
+  cd $SIFDECODE
+  echo "5" > sifdecode.opts   # PC
+  echo "1" >> sifdecode.opts  # Windows
+  echo "3" >> sifdecode.opts  # gfortran
+fi
+echo "nny" >> sifdecode.opts
+./install_sifdecode < sifdecode.opts
+
+# build CUTEst
+cd $CUTEST
+if [[ "${target}" == *-linux* || "${target}" == *-freebsd* ]]; then
+  echo "5" > cutest.opts   # PC
+  echo "2" >> cutest.opts  # Linux
+  echo "5" >> cutest.opts  # gfortran
+  echo "2" >> cutest.opts  # build all tools except Matlab
+  echo "7" >> cutest.opts  # gcc
+  export MYARCH=pc.lnx.gfo
+elif [[ "${target}" == *-apple* ]]; then
+  echo "13" > cutest.opts  # macOS
+  echo "2" >> cutest.opts  # gfortran
+  echo "2" >> cutest.opts  # build all tools except Matlab
+  echo "5" >> cutest.opts  # gcc
+  export MYARCH=mac64.osx.gfo
+elif [[ "${target}" == *-mingw* ]]; then
+  echo "5" > cutest.opts   # PC
+  echo "1" >> cutest.opts  # Windows
+  echo "3" >> cutest.opts  # gfortran
+  echo "2" >> cutest.opts  # build all tools except Matlab
+  echo "5" >> cutest.opts  # gcc
+  export MYARCH=pc.mgw.gfo
+fi
+echo "nnydy" >> cutest.opts
+./install_cutest < cutest.opts
+
+# build shared libs
+all_load="--whole-archive"
+noall_load="--no-whole-archive"
+extra=""
+if [[ "${target}" == *-apple-* ]]; then
+    all_load="-all_load"
+    noall_load="-noall_load"
+    extra="-Wl,-undefined -Wl,dynamic_lookup -headerpad_max_install_names"
+fi
+cd $CUTEST/objects/$MYARCH/double
+echo "gfortran -fPIC -shared ${extra} -Wl,${all_load} libcutest.a -Wl,${noall_load} -o libcutest_double.${dlext}"
+gfortran -fPIC -shared ${extra} -Wl,${all_load} libcutest.a -Wl,${noall_load} -o libcutest_double.${dlext}
+cd $CUTEST/objects/$MYARCH/single
+gfortran -fPIC -shared ${extra} -Wl,${all_load} libcutest.a -Wl,${noall_load} -o libcutest_single.${dlext}
+
+ln -s $SIFDECODE/bin/sifdecoder ${bindir}/
+ln -s $SIFDECODE/objects/$MYARCH/double/slct ${bindir}/
+ln -s $SIFDECODE/objects/$MYARCH/double/clsf ${bindir}/
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = [
+ Linux(:i686, libc=:glibc),
+ Linux(:x86_64, libc=:glibc),
+ Linux(:aarch64, libc=:glibc),
+ Linux(:armv7l, libc=:glibc, call_abi=:eabihf),
+ Linux(:powerpc64le, libc=:glibc),
+ Linux(:i686, libc=:musl),
+ Linux(:x86_64, libc=:musl),
+ Linux(:aarch64, libc=:musl),
+ Linux(:armv7l, libc=:musl, call_abi=:eabihf),
+ MacOS(:x86_64),
+ FreeBSD(:x86_64),
+ # Windows(:i686),  # can't build shared libs because Windows imposes all symbols to be defined
```

Comment:
https://github.com/JuliaPackaging/Yggdrasil/wiki/Build-Tricks#libtool-refuses-to-build-shared-library-because-of-undefined-symbols might help

Reply:
Thanks! I'll have a go at it.

Reply:
Unfortunately, this software isn't ready for libtool. I'm creating shared libs by hand from static libs.

Reply:
Well, the important thing is to pass -no-undefined to the linker in some way

Reply:
Adding `-Wl,-no-undefined` to the gfortran invocation for Windows should be sufficient. I'll probably need to reorganise the FAQ to make it clear that this isn't specific to libtools (that's just a way the problem can manifest) and the general solution is to pass a flag to the linker

Reply:
I was experimenting with that. The flag simply seems to be ignored and the linker keeps complaining. From `ld --help` it seems that `--no-undefined` is the opposite of what we want though. I also experimented with `-Wl,--unresolved-symbols=ignore-all` and `-Wl,--allow-shlib-undefined` but I keep getting the same complaints.

```
# ld --help
...
  --no-undefined              Do not allow unresolved references in object files
  --allow-shlib-undefined     Allow unresolved references in shared libraries
...
  --unresolved-symbols=<method>
                              How to handle unresolved symbols.  <method> is:
                                ignore-all, report-all, ignore-in-object-files,
                                ignore-in-shared-libs
```

Reply:
Ok, now that I've compiled locally I understand the error better:
```
sandbox:${WORKSPACE}/destdir/libexec/CUTEst-2.0.3/objects/pc.mgw.gfo/double # gfortran -fPIC -shared ${extra} -Wl,${all_load} libcutest.a -Wl,${noall_load} -o libcutest_double.${dlext}
libcutest.a(ufn.o): In function `cutest_ufn_threadsafe_':
/workspace/destdir/libexec/CUTEst-2.0.3/objects/pc.mgw.gfo/double/ufn.f90:133: undefined reference to `elfun_'
/workspace/destdir/libexec/CUTEst-2.0.3/objects/pc.mgw.gfo/double/ufn.f90:170: undefined reference to `group_'
libcutest.a(ugr.o): In function `cutest_ugr_threadsafe_':
/workspace/destdir/libexec/CUTEst-2.0.3/objects/pc.mgw.gfo/double/ugr.f90:126: undefined reference to `elfun_'
/workspace/destdir/libexec/CUTEst-2.0.3/objects/pc.mgw.gfo/double/ugr.f90:136: undefined reference to `elfun_'
/workspace/destdir/libexec/CUTEst-2.0.3/objects/pc.mgw.gfo/double/ugr.f90:168: undefined reference to `group_'
/workspace/destdir/libexec/CUTEst-2.0.3/objects/pc.mgw.gfo/double/ugr.f90:181: undefined reference to `range_'
[...]
```
Where are these function supposed to be defined?  Do you want to invest some time to fix the Windows build or you don't need it?

Reply:
Those functions will appear when the user interacts with the package. They will be compiled in a separate shared lib that will be linked against those provided in this package. That works fine on macOS and Linux but not Windows.

I'd like to get Windows working eventually, but it's not essential to have it working at this point.